cpp/src/arrow/util/decimal_internal.h

```diff
@@ -451,6 +451,8 @@ struct RealTraits<float> {
   static constexpr int kMantissaBits = 24;
   // ceil(log10(2 ^ kMantissaBits))
   static constexpr int kMantissaDigits = 8;
+  // Integers between zero and kMaxPreciseInteger can be precisely represented
+  static constexpr uint64_t kMaxPreciseInteger = 1ULL << kMantissaBits;
```

Comment:
Need to subtract one, no?
```suggestion
  static constexpr uint64_t kMaxPreciseInteger = (1ULL << kMantissaBits) - 1;
```